x-pack/filebeat/input/s3/collector.go

```diff
@@ -0,0 +1,635 @@
+// you may not use this file except in compliance with the Elastic License.
+
+package s3
+
+import (
+	"bufio"
+	"compress/gzip"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	awssdk "github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/awserr"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/s3/s3iface"
+	"github.com/aws/aws-sdk-go-v2/service/sqs"
+	"github.com/aws/aws-sdk-go-v2/service/sqs/sqsiface"
+	"github.com/pkg/errors"
+
+	"github.com/elastic/beats/v7/libbeat/beat"
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/logp"
+)
+
+type s3Collector struct {
+	cancellation context.Context
+	logger       *logp.Logger
+
+	config            *config
+	visibilityTimeout int64
+
+	sqs       sqsiface.ClientAPI
+	s3        s3iface.ClientAPI
+	publisher beat.Client
+}
+
+type s3Info struct {
+	name                     string
+	key                      string
+	region                   string
+	arn                      string
+	expandEventListFromField string
+}
+
+type bucket struct {
+	Name string `json:"name"`
+	Arn  string `json:"arn"`
+}
+
+type object struct {
+	Key string `json:"key"`
+}
+
+type s3BucketObject struct {
+	bucket `json:"bucket"`
+	object `json:"object"`
+}
+
+type sqsMessage struct {
+	Records []struct {
+		EventSource string         `json:"eventSource"`
+		AwsRegion   string         `json:"awsRegion"`
+		EventName   string         `json:"eventName"`
+		S3          s3BucketObject `json:"s3"`
+	} `json:"Records"`
+}
+
+type s3Context struct {
+	mux  sync.Mutex
+	refs int
+	err  error // first error witnessed or multi error
+	errC chan error
+}
+
+var (
+	// The maximum number of messages to return. Amazon SQS never returns more messages
+	// than this value (however, fewer messages might be returned).
+	maxNumberOfMessage int64 = 10
+
+	// The duration (in seconds) for which the call waits for a message to arrive
+	// in the queue before returning. If a message is available, the call returns
+	// sooner than WaitTimeSeconds. If no messages are available and the wait time
+	// expires, the call returns successfully with an empty list of messages.
+	waitTimeSecond int64 = 10
+)
+
+func (c *s3Collector) run() {
+	defer c.logger.Infof("s3 input worker for '%v' has stopped.", c.config.QueueURL)
+	c.logger.Infof("s3 input worker has started. with queueURL: %v", c.config.QueueURL)
+	fmt.Println("c.cancellation.Err() = ", c.cancellation.Err())
+	for c.cancellation.Err() == nil {
+		// receive messages from sqs
+		output, err := c.receiveMessage(c.sqs, c.visibilityTimeout)
+		if err != nil {
+			if awsErr, ok := err.(awserr.Error); ok && awsErr.Code() == awssdk.ErrCodeRequestCanceled {
+				continue
+			}
+			c.logger.Error("SQS ReceiveMessageRequest failed: ", err)
+			time.Sleep(time.Duration(waitTimeSecond) * time.Second)
```

Comment:
Why are you using `Sleep`? Why not use the backoff functionality provided by libbeat in `libbeat/common/backoff`?

Reply:
Not sure why this is even there 😂 `waitTimeSecond` is used in `SQS.ReceiveMessageRequest` in `ReceiveMessage` function. I will remove this `time.Sleep` here.